specs/data_structures.md

```diff
@@ -406,16 +416,153 @@ Wrapped transactions include additional metadata by the block proposer that is c
+| ------------ | ------------------- | ----------------------------------- |
+| `type`       | `TransactionType`   | Must be `TransactionType.Transfer`. |
+| `amount`     | `uint64`            |                                     |
+| `to`         | [Address](#address) |                                     |
+| `maxFeeRate` | `uint64`            |                                     |
+| `nonce`      | `uint64`            |                                     |
+
+Transfers `amount` coins to `to`.
+
+#### SignedTransactionData: PayForMessage
+
+| name                     | type                                       | description                                                  |
+| ------------------------ | ------------------------------------------ | ------------------------------------------------------------ |
+| `type`                   | `TransactionType`                          | Must be `TransactionType.PayForMessage`.                     |
+| `maxFeeRate`             | `uint64`                                   |                                                              |
+| `nonce`                  | `uint64`                                   |                                                              |
+| `messageNamespaceID`     | [`NamespaceID`](consensus.md#type-aliases) | Namespace ID of message this transaction pays a fee for.     |
+| `messageSize`            | `uint64`                                   | Size of message this transaction pays a fee for, in `byte`s. |
+| `messageShareCommitment` | [HashDigest](#hashdigest)                  | Commitment to message shares (details below).                |
+
+Pays for the inclusion of a [message](#message) in the same block.
+
+The commitment to message shares `messageShareCommitment` is a [Merkle root](#binary-merkle-tree) of message share roots. Each message share root is [a subtree root in a row NMT](#arranging-available-data-into-shares). For rationale, see [rationale doc](../rationale/message_block_layout.md).
+
+#### SignedTransactionData: PayForPadding
+
+| name                 | type                                       | description                                                  |
+| -------------------- | ------------------------------------------ | ------------------------------------------------------------ |
+| `type`               | `TransactionType`                          | Must be `TransactionType.PayForPadding`.                     |
+| `messageNamespaceID` | [`NamespaceID`](consensus.md#type-aliases) | Namespace ID of padding this transaction pays a fee for.     |
+| `messageSize`        | `uint64`                                   | Size of padding this transaction pays a fee for, in `byte`s. |
+
+Pays for the inclusion of a padding shares in the same block. Padding shares are used between real messages that are not tightly packed. For rationale, see [rationale doc](../rationale/message_block_layout.md).
```

Comment:
Will users send two tx then? one for the inclusion of the message and an optional extra one for padding only?

Reply:
The transactions here are based on the staking messages of the Cosmos SDK: https://docs.cosmos.network/master/modules/staking/03_messages.html. Some features are removed (one-transaction re-delagation, for example, since it adds protocol complexity that can be abstracted away at the wallet level).

Reply:
Not necessarily (i.e. no, but subject to change). Interstitial padding should be paid (or, at least, _claimed_) by the block proposer IMO.

Reply:
Clarified in c001036.